cdap-kms/src/main/java/co/cask/cdap/security/store/KMSSecureStore.java

```diff
@@ -0,0 +1,160 @@
+    this.conf = conf;
+    try {
+      URI providerUri = new URI(conf.get(KeyProviderFactory.KEY_PROVIDER_PATH));
+      provider = KMSClientProvider.Factory.get(providerUri, conf);
+    } catch (URISyntaxException e) {
+      throw new URISyntaxException("Secure store could not be loaded. The value for hadoop.security.key.provider.path" +
+                                     "in core-site.xml is not a valid URI.", e.getReason());
+    } catch (IOException e) {
+      throw new IOException("Secure store could not be loaded. KMS KeyProvider failed to initialize", e);
+    }
+    LOG.info("Secure Store initialized successfully.");
+  }
+
+  /**
+   * Stores an element in the secure store. The key is stored as namespace:name in the backing store,
+   * assuming ":" to be the separator.
+   * @param namespace The namespace this key belongs to.
+   * @param name Name of the element to store.
+   * @param data The data that needs to be securely stored.
+   * @param description User provided description of the entry.
+   * @param properties Metadata associated with the data
+   * @throws IOException If it failed to store the key in the store.
+   */
+  @Override
+  public void put(String namespace, String name, byte[] data, String description, Map<String, String> properties)
+    throws IOException {
+    KeyProvider.Options options = new KeyProvider.Options(conf);
+    options.setDescription(description);
+    options.setAttributes(properties);
+    options.setBitLength(data.length * Byte.SIZE);
+    String keyName = getKeyName(namespace, name);
+    try {
+      provider.createKey(keyName, data, options);
+    } catch (IOException e) {
+      throw new IOException("Failed to store the key. " + name + " under namespace " + namespace, e);
+    }
+  }
+
+  /**
+   * Deletes the element with the given name.
+   * @param namespace The namespace this key belongs to.
+   * @param name Name of the element to be deleted.
+   */
+  @Override
+  public void delete(String namespace, String name) throws IOException {
+    try {
+      provider.deleteKey(getKeyName(namespace, name));
+    } catch (IOException e) {
+      throw new IOException("Failed to delete the key. " + name + " under namespace " + namespace, e);
+    }
+  }
+
+  /**
+   * List of all the entries in the secure store.
+   * @return A list of {@link SecureStoreMetadata} objects representing the data stored in the store.
+   * @param namespace The namespace this key belongs to.
+   */
+  @Override
+  public List<SecureStoreMetadata> list(String namespace) throws IOException {
+    String prefix = namespace + NAME_SEPARATOR;
+    List<String> keysInNamespace = new ArrayList<>();
+    KeyProvider.Metadata[] metadatas;
+    try {
+      for (String key : provider.getKeys()) {
+        if (key.startsWith(prefix)) {
+          keysInNamespace.add(key);
+        }
+      }
+      metadatas = provider.getKeysMetadata(keysInNamespace.toArray(new String[0]));
```

Comment:
When using `toArray`, prefer to pass in an array of the right size, especially you know about the size in this case.